artifact/gem5art/artifact/_artifactdb.py

```diff
@@ -89,10 +149,47 @@ def downloadFile(self, key: UUID, path: str) -> None:
         with open(path, 'wb') as f:
             self.fs.download_to_stream(key, f)
 
+
+    def searchByName(self, name: str, limit: int) -> Iterable[Dict[str, Any]]:
+        """Returns an iterable of all artifacts in the database that match
+        some name."""
+        for d in self.artifacts.find({'name': name}, limit=limit):
+            yield d
+
+    def searchByType(self, typ: str, limit: int) -> Iterable[Dict[str, Any]]:
+        """Returns an iterable of all artifacts in the database that match
+        some type."""
+        for d in self.artifacts.find({'type':typ}, limit=limit):
+            yield d
+
+    def searchByNameType(self, name: str, typ: str, limit: int) -> Iterable[Dict[str, Any]]:
+        """Returns an iterable of all artifacts in the database that match
+        some name and type."""
+        for d in self.artifacts.find({'type':typ, 'name': name}, limit=limit):
+            yield d
+
+    def searchByLikeNameType(self, name: str, typ: str, limit: int) -> Iterable[Dict[str, Any]]:
+        """Returns an iterable of all artifacts in the database that match
+        some type and a regex name."""
+
+        data = self.artifacts.find({'type':typ,
+                                    'name': {'$regex': '{}'.format(name)}
+                                   },
+                                   limit=limit)
+        for d in data:
+            yield d
+
+_db = None
+
 def getDBConnection() -> ArtifactDB:
     """Returns the database connection
 
     Eventually, this should likely read from a config file to get the database
     information. However, for now, we'll use mongodb defaults
     """
-    return ArtifactDB()
+    global _db
```

Comment:
@powerjg : I have a confusion here: so, as far as I understand getDBConnection() will serve as the connection point with any kind of database e.g. for a mock/fake database (assuming that a class exists for it), we have to initiate it here and let _db point to it . I wonder why does _db has to be global?

Reply:
We want `getDBConnection` to return a singleton object. At any time, there should only be one DB connection. This is one way to implement the singleton design pattern in Python.

Right now, it is always a MongoDB. In the future, we can add parameters to this function to create different kinds of databases. However, after the first call to the function, all other calls should return the same instance of the database connection (i.e., a singleton)

Reply:
makes sense, thanks!